sklearn/tree/_splitter.pyx

```diff
@@ -550,199 +507,296 @@ cdef void heapsort(DTYPE_t* Xf, SIZE_t* samples, SIZE_t n) nogil:
-            # - [f_i:n_features[ holds features that have been drawn
-            #   and aren't constant.
-
-            # Draw a feature at random
-            f_j = rand_int(n_drawn_constants, f_i - n_found_constants,
-                           random_state)
-
-            if f_j < n_known_constants:
-                # f_j in the interval [n_drawn_constants, n_known_constants[
-                features[n_drawn_constants], features[f_j] = features[f_j], features[n_drawn_constants]
-                n_drawn_constants += 1
-                continue
-
-            # f_j in the interval [n_known_constants, f_i - n_found_constants[
-            f_j += n_found_constants
-            # f_j in the interval [n_total_constants, f_i[
-
-            current.feature = features[f_j]
-
-            # Find min, max
-            min_feature_value = self.X[samples[start], current.feature]
-            max_feature_value = min_feature_value
-            Xf[start] = min_feature_value
-
-            for p in range(start + 1, end):
-                current_feature_value = self.X[samples[p], current.feature]
-                Xf[p] = current_feature_value
-
-                if current_feature_value < min_feature_value:
-                    min_feature_value = current_feature_value
-                elif current_feature_value > max_feature_value:
-                    max_feature_value = current_feature_value
-
-            if max_feature_value <= min_feature_value + FEATURE_THRESHOLD:
-                features[f_j], features[n_total_constants] = features[n_total_constants], current.feature
-
-                n_found_constants += 1
-                n_total_constants += 1
-                continue
-
-            f_i -= 1
-            features[f_i], features[f_j] = features[f_j], features[f_i]
-
-            # Draw a random threshold
-            current.threshold = rand_uniform(min_feature_value,
-                                             max_feature_value,
-                                             random_state)
-
-            if current.threshold == max_feature_value:
-                current.threshold = min_feature_value
 
-            # Partition
-            p, partition_end = start, end
-            while p < partition_end:
-                if Xf[p] <= current.threshold:
-                    p += 1
-                else:
-                    partition_end -= 1
+        while p[0] + 1 < self.end and Xf[p[0] + 1] <= Xf[p[0]] + FEATURE_THRESHOLD:
+            p[0] += 1
 
-                    Xf[p], Xf[partition_end] = Xf[partition_end], Xf[p]
-                    samples[p], samples[partition_end] = samples[partition_end], samples[p]
+        # (p + 1 >= end) or (Xf[p + 1] > X[p])
+        p[0] += 1
+        # (p >= end) or (X[p, current.feature] > X[p - 1, current.feature])
+        p_prev[0] = p[0] - 1
 
-            current.pos = partition_end
+    cdef inline SIZE_t parition_samples(self, double current_threshold) nogil:
```

Comment:
```suggestion
    cdef inline SIZE_t partition_samples(self, double current_threshold) nogil:
```

Reply:
Would it make sense to actually name it `partition_samples_random` since there is also a `partition_samples_best`?

Reply:
I renamed this to `partition_samples_final`. It is called by both the `best` and `random` splitter when the split point is finalized.